src/main/java/com/snowflake/kafka/connector/internal/streaming/BufferedTopicPartitionChannel.java

```diff
@@ -517,17 +515,25 @@ public InsertRowsResponse insertRecords(StreamingBuffer streamingBufferToInsert)
           streamingBufferToInsert,
           response.hasErrors(),
           response.needToResetOffset());
-      if (response.hasErrors()) {
-        handleInsertRowsFailures(
-            response.getInsertErrors(), streamingBufferToInsert.getSinkRecords());
-      }
 
       // Due to schema evolution, we may need to reopen the channel and reset the offset in kafka
       // since it's possible that not all rows are ingested
       if (response.needToResetOffset()) {
         streamingApiFallbackSupplier(
             StreamingApiFallbackInvoker.INSERT_ROWS_SCHEMA_EVOLUTION_FALLBACK);
       }
+      // If there are errors other than schema mismatch, we need to handle them and reinsert the
+      // good rows
+      if (response.hasErrors()) {
+        handleInsertRowsFailures(
+            response.getInsertErrors(), streamingBufferToInsert.getSinkRecords());
+        insertRecords(
```

Comment:
do we need to reset the offset if we put here the schema evolution records as well?
@sfc-gh-tzhang

Reply:
I don't think so, why we need to do that?

Reply:
we are setting this flag needToResetOffset to be true in case of schema evolution error. I'm not sure what it exactly does and if we need this if we now reinsert the rows by rebuild the buffer with rebuildBufferWithoutErrorRows

Reply:
if you see how needToResetOffset will be used, when it's true, we will reopen the channel and restart from the latest committed offset token